batchflow/models/tf/xception.py

```diff
@@ -0,0 +1,214 @@
+                for i in range(exit_stages):
+                    args = {**kwargs, **exit, **unpack_args(exit, i, exit_stages)}
+                    x = cls.block(x, name='block-'+str(i), **args)
+
+            with tf.variable_scope('entry/group-'+str(entry_stages)):
+                x = tf.identity(x, name='output')
+        return x
+
+    @classmethod
+    def block(cls, inputs, filters, combine_type='sum', name='block', **kwargs):
+        """ Basic building block of the architecture.
+        For details see figure 5 in the article.
+
+        Parameters
+        ----------
+        filters : sequence of 3 ints
+            Number of feature maps in each layer
+        """
+        strides = (1, kwargs.pop('strides', 1), 1)
+        x = inputs
+
+        with tf.variable_scope(name):
+            # Prepare skip-connection
+            if combine_type == 'sum':
+                shortcut = inputs
+            elif combine_type == 'conv':
+                shortcut = conv_block(inputs, 'cn', filters[-1], kernel_size=1,
+                                      strides=strides[1], name='shortcut')
+
+            # Three consecutive separable blocks
+            for i, filter_ in enumerate(filters):
+                x = cls.separable_block(x, filter_, strides=strides[i],
+                                        name='separable_conv-{}'.format(i), **kwargs)
+
+            outputs = tf.add_n([x, shortcut])
+        return outputs
+
+    @classmethod
+    def separable_block(cls, inputs, filters, kernel_size=3, strides=1, rate=1,
+                        depth_activation=False, name='separable_block', **kwargs):
+        """ Separable convolution, followed by pointwise convolution with batch-normalization in-between. """
+        layout = 'nacna' if depth_activation else 'ncn'
+        data_format = kwargs.get('data_format')
+
+        with tf.variable_scope(name):
+            x = depthwise_conv(inputs, kernel_size=kernel_size, strides=strides, dilation_rate=rate,
+                               data_format=data_format, padding='same', name='depthwise')
+            x = conv_block(x, layout, filters=filters, kernel_size=1, **kwargs)
+        return x
+
+
+    @classmethod
+    def make_encoder(cls, inputs, name='encoder', **kwargs):
+        """ Build the body and return the last tensors of each spatial resolution.
+
+        Parameters
+        ----------
+        inputs : tf.Tensor
+            input tensor
+        name : str
+            scope name
+        kwargs : dict
+            body params
+        """
+        steps = 1 + cls.get('entry/num_stages', config=cls.fill_params('body', **kwargs))
+
+        with tf.variable_scope(name):
+            x = cls.body(inputs, name='body', **kwargs)
+
+            scope = tf.get_default_graph().get_name_scope()
+            encoder_tensors = [inputs]
+            for i in range(steps):
+                tensor_name = scope + '/body/entry/group-'+str(i) + '/output:0'
```

Comment:
Let's create a default TFModel's `make_encoder` which build body and the gets tensors for a given regexp